### PR TITLE
deploy: remove deployer pods on cancellation

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/openshift/origin/issues/8439
Fixes https://github.com/openshift/origin/issues/8500
Fixes https://github.com/openshift/origin/issues/9289

@smarterclayton @ironcladlou @mfojtik PTAL

### CODE REVIEW

**Comment by kargakis** [comment]
Opened https://github.com/openshift/origin/issues/9296 which is related

**Comment by smarterclayton** [comment]
[test]

**Comment by kargakis** [comment]
```
• Failure [24.149 seconds]
deploymentconfigs
/data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:309
  when run iteratively
  /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:131
    should only deploy the last deployment [Conformance] [It]
    /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:130

    Expected error:
        <*errors.errorString | 0xc8201c1630>: {
            s: "found multiple running deployments: [deployment-simple-1 deployment-simple-2 deployment-simple-3 deployment-simple-4]",
        }
        found multiple running deployments: [deployment-simple-1 deployment-simple-2 deployment-simple-3 deployment-simple-4]
    not to have occurred

    /data/src/github.com/openshift/origin/test/extended/deployments/deployments.go:119
```
The failure is real but I think it has to do with the way the deployer pods are listed in one shot in the extended utilities for the deployments. Looking...

**Comment by kargakis** [comment]
Worth noting that I don't see anywhere in the output the defered function that dumps the dc and its rcs and deployer pods.

**Comment by kargakis** [comment]
Ok, so we cannot just transition the deployment to Failed when we delete the deployers because the deployment config controller will notice it and create a new deployment maybe before the previous deployer pod is deleted. We can transition to Failed once we make sure the deployer pod does not exist. I think we will need a pod cache to requeue on DELETE events (we will need to requeue on every event for https://github.com/openshift/origin/issues/9296 anyway).

**Comment by kargakis** [comment]
Green on the extended test, broken unit test

File: pkg/deploy/controller/deployment/controller.go [hunk 1]
```diff
@@ -126,29 +128,34 @@ func (c *DeploymentController) Handle(deployment *kapi.ReplicationController) er
...
-		// The deployment will be marked as Failed by the deployer pod controller
-		// when the deployer pod failure state is picked up.
-		// Then, the deployment config controller will scale down the failed deployment
-		// and scale back up the last successful completed deployment.
-		if deployutil.IsDeploymentCancelled(deployment) {
-			if err := c.cancelDeployerPods(deployment); err != nil {
-				return err
+		case err != nil:
+			// We'll try again later on resync. Continue to process cancellations.
+			glog.V(2).Infof("Error getting deployer pod %s for deployment %s: %#v", deployerPodName, deployutil.LabelForDeployment(deployment), err)
```
Comment (smarterclayton) [inline pkg/deploy/controller/deployment/controller.go]: v(4) probably.

**Comment by smarterclayton** [comment]
Changes look ok

**Comment by kargakis** [comment]
yum flake, repushed to fix the glog level

**Comment by smarterclayton** [comment]
Lgtm [merge]

**Pull request merged**